File: display/preview.lua
-- The preview keys, the workstation stand-in for the bus. On a cluster the idle
-- pod's sidecar forwards the Player's retained status and the reveal, and there
-- is no keyboard. On a workstation local/idle sets IDLE_PREVIEW=1 and this
-- module binds keys that build the same JSON the sidecar sends and hand it to
-- the same handler, so the ramps, the dim, and the pulse can be seen before a
-- release. Nothing here runs with the variable unset.
local utils = require("mp.utils")
local theme = require("theme")

local preview = {}

-- Whether enable has run. The legend draws only then, so the module can be
-- required unconditionally and still draw nothing on a cluster.
local enabled = false

-- The parts come from the environment seed, which carries names and no kinds.
-- identity owns that seed's format and exports the splitter for it. The
-- preview calls the last part the remote, because the parts read in the
-- order the operator lists them and a controller is the part a person adds
-- last. That part is the one whose presence the d key toggles.
local identity = require("identity")

local kind = "Idle"
local connected = true

-- Whether the fake status marks the remote focused. The f key sets it and
-- pulses, the way a live mark landing here does, and the g key clears it,
-- the way a cycle onto another unit does.
local focused = false

-- Whether the s key last put the screen to sleep, so the one key plays
-- the two edges of the fade in turn.
local asleep = false

local function status_json()
  local names = identity.split_lines(os.getenv("IDLE_PLAYER_COMPONENTS"))
  local components = {}
  for index, name in ipairs(names) do
    local component = { name = name, kind = "sink" }
    if index == #names then
      component.kind = "remote"
      component.connected = connected
      if focused then
        component.focused = true
      end
    end
    components[#components + 1] = component
  end
  local status = {
    displayName = os.getenv("IDLE_PLAYER_NAME"),
    activity = kind,
    components = components,
  }
  -- The Play block appears only while a Play starts or runs, the way the
  -- operator publishes it.
  if kind ~= "Idle" then
    status.play = { name = "sailing", title = "Sailing" }
  end
  return utils.format_json(status)
end

-- enable binds the keys. send_status takes the status as a JSON string and
-- send_revealed takes no argument, so both are the handlers the script messages
-- call and the preview exercises the same code.
function preview.enable(send_status, send_revealed, send_sleep, send_wake)
  local function publish(activity)
    kind = activity
    send_status(status_json())
  end

  mp.add_key_binding("p", "liken-preview-starting", function()
    publish("Starting")
  end)
  mp.add_key_binding("o", "liken-preview-playing", function()
    publish("Playing")
  end)
  -- i plays the end of a film: the status returns to Idle and the sidecar
  -- reports the idle surface back in view, in that order.
  mp.add_key_binding("i", "liken-preview-idle", function()
    publish("Idle")
    send_revealed()
  end)
  mp.add_key_binding("d", "liken-preview-presence", function()
    connected = not connected
    send_status(status_json())
  end)
  -- f plays a live focus message naming this unit: the sidecar sends the
  -- pulse and the operator's status follows with the focused part. A first
  -- press draws the hexagon and beats it, and every later press beats it
  -- again, which is the cycle press that wraps onto the unit already
  -- focused. The index is 0 because the preview's status lists one remote.
  mp.add_key_binding("f", "liken-preview-focus", function()
    focused = true
    send_status(status_json())
    mp.commandv("script-message", "focus-pulse", "0")
  end)
  -- g plays the focus cycling away: the status stops marking the remote,
  -- the hexagon goes, and nothing beats.
  mp.add_key_binding("g", "liken-preview-unfocus", function()
    focused = false
    send_status(status_json())
  end)
  -- s plays the two edges of the fade in turn, the quiet window that
  -- runs out and the press that follows, through the same handlers the
  -- two script messages call.
  mp.add_key_binding("s", "liken-preview-sleep", function()
    asleep = not asleep
    if asleep then
      send_sleep()
    else
      send_wake()
    end
  end)

  -- The volume keys play the idle sidecar's part: on a cluster a press
  -- publishes the unit's next state, and the sidecar sets the properties
  -- and sends volume-changed. Here the key does both, so the indicator
  -- shows on the idle surface the way a press shows it. 9, 0, and m are
  -- mpv's own volume keys.
  local function press_volume(step)
    local level = mp.get_property_number("volume") or 100
    level = math.max(0, math.min(100, level + step))
    mp.set_property_number("volume", level)
    mp.commandv("script-message", "volume-changed")
  end
  mp.add_key_binding("9", "liken-preview-volume-down", function()
    press_volume(-5)
  end, { repeatable = true })
  mp.add_key_binding("0", "liken-preview-volume-up", function()
    press_volume(5)
  end, { repeatable = true })
  mp.add_key_binding("m", "liken-preview-mute", function()
    mp.set_property_bool("mute", not mp.get_property_bool("mute"))
    mp.commandv("script-message", "volume-changed")
  end)

  mp.msg.info("liken display preview keys: p starting, o playing, i idle, d presence, f focus, g unfocus, s sleep, 9/0 volume, m mute")
  enabled = true
end

-- The legend, one dim line at the bottom right, so the keys read on the screen
-- itself and not only in this file. It draws only after enable ran, which only
-- local/idle causes, so an idle pod on a cluster draws no legend. The an3
-- alignment anchors the line at its bottom-right, mirroring the identity block's
-- bottom-left across the screen.
function preview.draw()
  if not enabled then
    return nil
  end
  local legend = "p play starts    o playing    i film ends    d presence    f focus    g unfocus    s sleep    9/0 volume    m mute    q quit"
  return theme.text(
    theme.canvas.w - theme.margin.x, theme.canvas.h - theme.margin.y,
    legend, theme.type.tiny, theme.color.muted, 3, theme.alpha.dim
  )
end

return preview

File: idle/src/idle/preview.rs
//! The preview keys, the workstation stand-in for the bus, and the legend that
//! names them on the screen itself.
//!
//! On a cluster the idle command pod publishes the `Player`'s status, its level,
//! and the screen's moments, and the pod has no keyboard. On a workstation
//! `local/idle` sets `IDLE_PREVIEW=1` and the keys below build the same
//! [`Moment`] values those topics carry. The client folds each one through
//! `Client::receive`, the call the bus reader's own messages take, so a person
//! plays every ramp, every dim, and every beat against the real handlers and
//! not against a second set written for a workstation.

use iced_winit::core::Point;

use super::{Frame, Layout};
use crate::look;
use media_screen::Moment;
use media_screen::status::{Activity, Component, Play, Status};
use media_screen::volume::{MIN_LEVEL, UNITY_LEVEL, Volume};

/// The kind the preview gives every part but the last. A screen and a set of
/// speakers both draw at full brightness with no presence, so one word covers
/// both and the last part alone carries the presence the `d` key toggles.
const SINK: &str = "sink";

/// The `Play` the preview names while one starts or runs. The operator
/// publishes the object's name and the one line the screen draws, so the
/// preview states both.
const PLAY_NAME: &str = "sailing";
const PLAY_TITLE: &str = "Sailing";

/// The controller a focus moment names. A focus counts the controllers in the
/// order the status lists them, and the preview's status lists one, so the
/// index is always the first.
const REMOTE: usize = 0;

/// How far one volume key moves the level, out of the 100 that is unity: a
/// step big enough to read on the bar at a glance, and small enough that the
/// range takes twenty of them.
const VOLUME_STEP: i64 = 5;

/// The state behind the keys: what the fake status says about the unit now.
///
/// The bus holds this state on a cluster, and the client only reads it. With no
/// broker the presses have to hold it themselves, so each key changes one field
/// here and then states the whole unit again, the way the operator republishes
/// a whole status for one change.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Keys {
    /// The unit's name and its parts, from the same environment seed the
    /// operator sets on the idle container. The preview calls the last part the
    /// remote, because the parts read in the order the operator lists them and
    /// a controller is the part a person adds last.
    name: String,
    parts: Vec<String>,
    activity: Activity,
    /// Whether the remote is connected. The `d` key toggles it.
    connected: bool,
    /// Whether the status marks the remote focused. The `f` key sets it and
    /// beats the marker, the way a live mark landing here does, and the `g` key
    /// clears it, the way a cycle onto another unit does.
    focused: bool,
    /// Whether the `s` key last put the screen to sleep, so the one key plays
    /// the two edges of the fade in turn.
    asleep: bool,
    volume: Volume,
}

impl Keys {
    /// The keys for a unit the environment named. The seed carries names and no
    /// kinds, so every part reads as a sink until the last one, which is the
    /// remote.
    pub fn seeded(name: String, parts: Vec<String>) -> Self {
        Self {
            name,
            parts,
            activity: Activity::Idle,
            connected: true,
            focused: false,
            asleep: false,
            volume: Volume::default(),
        }
    }

    /// The messages one key press stands for, in the order the bus would carry
    /// them. A key this module does not bind carries nothing, so the arrow keys
    /// the harness also delivers change nothing here.
    pub fn press(&mut self, key: &str) -> Vec<Moment> {
        match key {
            "p" => vec![self.doing(Activity::Starting)],
            "o" => vec![self.doing(Activity::Playing)],
            // The end of a film: the status returns to `Idle` and the sidecar
            // reports the idle surface back in view, in that order.
            "i" => vec![self.doing(Activity::Idle), Moment::Present],
            "d" => {
                self.connected = !self.connected;
                vec![self.status()]
            }
            // A live focus message naming this unit: the status marks the
            // remote and the moment beats its marker. A first press draws the
            // marker and beats it, and every later press beats it again, which
            // is the cycle press that wraps onto the unit already focused.
            "f" => {
                self.focused = true;
                vec![self.status(), Moment::Focus { remote: REMOTE }]
            }
            // The focus cycling away. The status stops marking the remote, the
            // marker goes, and nothing beats.
            "g" => {
                self.focused = false;
                vec![self.status()]
            }
            "s" => {
                self.asleep = !self.asleep;
                vec![match self.asleep {
                    true => Moment::Sleep,
                    false => Moment::Wake,
                }]
            }
            "9" => vec![self.step(-VOLUME_STEP)],
            "0" => vec![self.step(VOLUME_STEP)],
            "m" => {
                self.volume.muted = !self.volume.muted;
                vec![self.level()]
            }
            _ => Vec::new(),
        }
    }

    /// Move the unit to one activity and state the whole status again.
    fn doing(&mut self, activity: Activity) -> Moment {
        self.activity = activity;
        self.status()
    }

    /// The status as the operator would publish it now. The `Play` block
    /// appears only while a `Play` starts or runs, which is the rule the
    /// operator writes it under.
    fn status(&self) -> Moment {
        let last = self.parts.len().saturating_sub(1);
        let components = self
            .parts
            .iter()
            .enumerate()
            .map(|(index, name)| {
                if index == last {
                    Component {
                        name: name.clone(),
                        kind: Component::REMOTE.to_string(),
                        connected: Some(self.connected),
                        focused: self.focused.then_some(true),
                    }
                } else {
                    Component {
                        name: name.clone(),
                        kind: SINK.to_string(),
                        connected: None,
                        focused: None,
                    }
                }
            })
            .collect();

        Moment::Status(Status {
            display_name: self.name.clone(),
            activity: self.activity,
            play: (self.activity != Activity::Idle).then(|| Play {
                name: PLAY_NAME.to_string(),
                title: PLAY_TITLE.to_string(),
            }),
            components,
        })
    }

    /// Move the level by one step, held inside the range the operator holds it
    /// in, and state it.
    fn step(&mut self, step: i64) -> Moment {
        self.volume.level = (self.volume.level + step).clamp(MIN_LEVEL, UNITY_LEVEL);
        self.level()
    }

    /// The level as a press. The first level of a bus session is the broker's
    /// retained catch-up, which a person did not press and which draws no
    /// indicator. A key is a press every time, so the indicator shows every
    /// time.
    fn level(&self) -> Moment {
        Moment::Level {
            volume: self.volume,
            pressed: true,
        }
    }
}

/// The legend, one key and what it does per pair, in the order the line reads
/// them. The screen names the keys itself and not only in this file.
const LEGEND: [(&str, &str); 10] = [
    ("p", "play starts"),
    ("o", "playing"),
    ("i", "film ends"),
    ("d", "presence"),
    ("f", "focus"),
    ("g", "unfocus"),
    ("s", "sleep"),
    ("9/0", "volume"),
    ("m", "mute"),
    // The one key of the legend this module does not bind. The harness reads
    // the quit key itself, and the legend names it beside the rest because a
    // person reading the line needs the way out.
    (crate::harness::QUIT, "quit"),
];

/// The gap between one pair and the next, wide enough that a reader takes the
/// key and its words as one.
const LEGEND_GAP: &str = "    ";

/// The legend as one line.
fn legend_line() -> String {
    LEGEND
        .iter()
        .map(|(key, does)| format!("{key} {does}"))
        .collect::<Vec<String>>()
        .join(LEGEND_GAP)
}

/// Draw the legend. The client calls this only where the preview keys are
/// bound, so a cluster's idle screen carries none.
///
/// The line hangs from its own bottom-right corner, which mirrors the identity
/// block's bottom-left across the screen.
pub fn legend(frame: &mut Frame, layout: &Layout, light: f32) {
    frame.fill_text(look::line(
        legend_line(),
        Point::new(layout.right(), layout.bottom()),
        look::Anchor::BottomRight,
        look::TINY,
        look::under(look::faded(look::muted(), look::DIM), light),
    ));
}

#[cfg(test)]
mod tests {
    use super::*;

    fn keys() -> Keys {
        Keys::seeded(
            "Studio Lab".to_string(),
            vec![
                "Portable Screen".to_string(),
                "Built-in Speakers".to_string(),
                "Studio Dualsense".to_string(),
            ],
        )
    }

    fn status(moment: &Moment) -> &Status {
        let Moment::Status(status) = moment else {
            panic!("the press states a status");
        };
        status
    }

    #[test]
    fn the_seed_names_the_unit_and_its_parts() {
        let pressed = keys().press("d");
        let status = status(&pressed[0]);

        assert_eq!(status.display_name, "Studio Lab");
        assert_eq!(
            status
                .components
                .iter()
                .map(|part| part.name.as_str())
                .collect::<Vec<_>>(),
            ["Portable Screen", "Built-in Speakers", "Studio Dualsense"]
        );
    }

    #[test]
    fn the_last_part_is_the_remote_and_every_other_part_is_a_sink() {
        let pressed = keys().press("d");
        let parts = &status(&pressed[0]).components;

        assert_eq!(parts[0].kind, SINK);
        assert_eq!(parts[0].connected, None);
        assert_eq!(parts[2].kind, Component::REMOTE);
        assert_eq!(parts[2].connected, Some(false));
    }

    #[test]
    fn a_unit_with_no_parts_states_none() {
        let mut keys = Keys::seeded("Studio Lab".to_string(), Vec::new());
        assert!(status(&keys.press("d")[0]).components.is_empty());
    }

    #[test]
    fn the_play_keys_state_the_three_activities() {
        let mut keys = keys();

        assert_eq!(status(&keys.press("p")[0]).activity, Activity::Starting);
        assert_eq!(status(&keys.press("o")[0]).activity, Activity::Playing);
        assert_eq!(status(&keys.press("i")[0]).activity, Activity::Idle);
    }

    #[test]
    fn a_play_names_a_title_and_a_unit_at_rest_names_none() {
        let mut keys = keys();

        let starting = keys.press("p");
        assert_eq!(
            status(&starting[0]).play,
            Some(Play {
                name: PLAY_NAME.to_string(),
                title: PLAY_TITLE.to_string(),
            })
        );

        let idle = keys.press("i");
        assert_eq!(status(&idle[0]).play, None);
    }

    #[test]
    fn the_end_of_a_film_returns_the_status_and_then_the_surface() {
        let pressed = keys().press("i");

        assert_eq!(pressed.len(), 2);
        assert_eq!(status(&pressed[0]).activity, Activity::Idle);
        assert_eq!(pressed[1], Moment::Present);
    }

    #[test]
    fn the_presence_key_disconnects_the_remote_and_connects_it_again() {
        let mut keys = keys();

        let gone = keys.press("d");
        assert_eq!(status(&gone[0]).components[2].connected, Some(false));

        let back = keys.press("d");
        assert_eq!(status(&back[0]).components[2].connected, Some(true));
    }

    #[test]
    fn focus_marks_the_remote_and_beats_its_marker() {
        let mut keys = keys();

        let landed = keys.press("f");
        assert_eq!(landed.len(), 2);
        assert_eq!(status(&landed[0]).components[2].focused, Some(true));
        assert_eq!(landed[1], Moment::Focus { remote: 0 });

        // The cycle press that wraps onto the unit already focused beats it
        // again.
        assert_eq!(keys.press("f")[1], Moment::Focus { remote: 0 });
    }

    #[test]
    fn focus_cycling_away_stops_marking_the_remote_and_beats_nothing() {
        let mut keys = keys();
        keys.press("f");

        let left = keys.press("g");

        assert_eq!(left.len(), 1);
        assert_eq!(status(&left[0]).components[2].focused, None);
    }

    #[test]
    fn the_sleep_key_plays_the_two_edges_of_the_fade_in_turn() {
        let mut keys = keys();

        assert_eq!(keys.press("s"), [Moment::Sleep], "the quiet window ran out");
        assert_eq!(keys.press("s"), [Moment::Wake], "a press woke the screen");
    }

    #[test]
    fn a_volume_key_steps_the_level_as_a_press() {
        let mut keys = keys();

        assert_eq!(
            keys.press("9"),
            [Moment::Level {
                volume: Volume {
                    level: UNITY_LEVEL - VOLUME_STEP,
                    muted: false,
                },
                pressed: true,
            }]
        );
        assert_eq!(
            keys.press("0"),
            [Moment::Level {
                volume: Volume {
                    level: UNITY_LEVEL,
                    muted: false,
                },
                pressed: true,
            }]
        );
    }

    #[test]
    fn the_level_stays_inside_the_range_the_operator_holds_it_in() {
        let mut keys = keys();
        for _ in 0..30 {
            keys.press("9");
        }
        assert_eq!(keys.volume.level, MIN_LEVEL);

        for _ in 0..30 {
            keys.press("0");
        }
        assert_eq!(keys.volume.level, UNITY_LEVEL);
    }

    #[test]
    fn the_mute_key_carries_the_muted_state_on_the_same_indicator() {
        let mut keys = keys();

        assert_eq!(
            keys.press("m"),
            [Moment::Level {
                volume: Volume {
                    level: UNITY_LEVEL,
                    muted: true,
                },
                pressed: true,
            }]
        );
        assert_eq!(
            keys.press("m"),
            [Moment::Level {
                volume: Volume {
                    level: UNITY_LEVEL,
                    muted: false,
                },
                pressed: true,
            }]
        );
    }

    #[test]
    fn a_key_this_module_does_not_bind_carries_nothing() {
        let mut keys = keys();
        assert!(keys.press("up").is_empty());
        assert!(keys.press("z").is_empty());
        assert_eq!(keys, self::keys());
    }

    #[test]
    fn the_preview_binds_every_key_the_legend_names_but_the_harness_key() {
        let mut keys = keys();
        let bound_to_nothing: Vec<&str> = LEGEND
            .iter()
            .flat_map(|(key, _)| key.split('/'))
            .filter(|key| keys.press(key).is_empty())
            .collect();

        assert_eq!(bound_to_nothing, [crate::harness::QUIT]);
    }

    #[test]
    fn the_legend_line_reads_each_key_beside_what_it_does() {
        assert_eq!(
            legend_line(),
            "p play starts    o playing    i film ends    d presence    f focus    \
             g unfocus    s sleep    9/0 volume    m mute    q quit"
        );
    }
}
